staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -260,20 +282,47 @@ func ServerPreferredResources(d DiscoveryInterface) ([]*metav1.APIResourceList,
 		return nil, err
 	}
 
-	result := []*metav1.APIResourceList{}
+	groupVersionResources := make(map[schema.GroupVersion]*metav1.APIResourceList)
 	failedGroups := make(map[schema.GroupVersion]error)
 
+	wg := &sync.WaitGroup{}
+	resultLock := &sync.Mutex{}
+	for _, apiGroup := range serverGroupList.Groups {
```

Comment:
Seems to have the exact method as above. We could make this as a common function, right?